### ISSUE TITLE
systemd-sysupdate: Query size when describing an update

### ISSUE THREAD

**Issue by AdrianVovk** [systemd/systemd#34710]
### Component

other

### Is your feature request related to a problem? Please describe

sysupdate and sysupdated don't report any download size information when enumerating an update. Graphical software centers need this information for display

### Describe the solution you'd like

sysupdate should somehow enumerate the download size from the download backend. Then return it, and pass it through sysupdated's `Describe()` method.

The one thing I'm pretty sure about is that sysupdate should calculate the total download size for the client, so the client doesn't need to traverse the `contents` object and sum up the sizes.

### Describe alternatives you've considered

:shrug:

### The systemd version you checked that didn't have the feature you are asking for

257

**Comment by AdrianVovk** [comment]
Discussing w/ the KDE OS folks about this, went for a walk, then came back with an idea:

We can probably just use appstream metadata for this. In appstream, you can encode the download size in the `<release>` tag. Maybe that's good enough, and sysupdate doesn't actually need to do anything?

On the other hand, would we have multiple metainfo files for each of our device-specific images? I suppose we would; if we list that we improved op6 camera support, that wouldn't make sense to put in the normal PC update's changelog. So we'd have per-board per-stream changelogs, which would permit us to put download size information in there too.

Thoughts? @alatiera @valentindavid

**Comment by shocklateboy92** [comment]
Also, I heard people were musing about the idea of doing some intelligent diffs to reduce the download size in the future.

If that's the case, it might make sense for `sysupdate` to tell clients the download size so features like that will be transparent to them.

**Comment by AdrianVovk** [comment]
> intelligent diffs to reduce the download size in the future

Ah boy that's not trivial. Some of the intelligent diffing schemes don't actually know exactly how much they need to download until they start, and even then others don't know until they're actually done. This is why Flatpak always lists sizes as "< X MB", for example.

Also, if we have #34584, getting `Content-Length:` from the http server isn't enough either. So essentially even basic downloads turn into the diffing case

**Comment by shocklateboy92** [comment]
> So essentially even basic downloads turn into the diffing case

Yep, sounds quite difficult. But that's all the more reason clients and CI servers shouldn't get involved.
If the thing that's doing the work calculates what needs to be done, it's more likely to be correct :grin:

**Comment by alatiera** [comment]
> We can probably just use appstream metadata for this. In appstream, you can encode the download size in the `<release>` tag. Maybe that's good enough, and sysupdate doesn't actually need to do anything?

Thing is usually the appstream files are part of the repo and release and they get modified with the new date, version etc, *before* you merge and tag.

**Comment by Justinzobel** [comment]
This would be great for users with data caps on their internet connection!
